File: python/slide_text_extractor.py
"""
SlideTextExtractor (§6, §6.2 ТЗ).

Вызывается из PHP-класса CasesBot\\Presentation\\SlideTextExtractor как подпроцесс.
Читает текст слайдов (заголовок, буллеты) через python-pptx — источник подсказок
для тегирования при индексации (Indexer). Печатает в stdout JSON-массив:
[{"slide_number": 1, "title": "...", "text": "...", "is_case": true}, ...]

Слайд считается кейсом, только если в заметках к слайду есть метка CASE_MARKER
(#кейс# по умолчанию) — так эксперт помечает слайды с кейсами прямо в PowerPoint:
метка видна только в заметках докладчика, не в самом слайде и не в режиме показа.

Использование: python slide_text_extractor.py <путь_к_pptx> [метка]
"""

import json
import sys

CASE_MARKER = "#кейс#"


def slide_title(slide, texts):
    title_shape = slide.shapes.title
    if title_shape is not None:
        title = title_shape.text_frame.text.strip()
        if title:
            return title

    return texts[0].splitlines()[0] if texts else ""


def notes_text(slide):
    if not slide.has_notes_slide:
        return ""
    return slide.notes_slide.notes_text_frame.text.strip()


def extract(path, case_marker):
    from pptx import Presentation

    presentation = Presentation(path)
    slides = []

    for index, slide in enumerate(presentation.slides, start=1):
        texts = []
        for shape in slide.shapes:
            if shape.has_text_frame:
                text = shape.text_frame.text.strip()
                if text:
                    texts.append(text)

        notes = notes_text(slide)

        slides.append({
            "slide_number": index,
            "title": slide_title(slide, texts),
            "text": "\n".join(texts),
            "is_case": case_marker.lower() in notes.lower(),
        })

    return slides


def main():
    if len(sys.argv) not in (2, 3):
        print("usage: slide_text_extractor.py <path.pptx> [case_marker]", file=sys.stderr)
        sys.exit(1)

    case_marker = sys.argv[2] if len(sys.argv) == 3 else CASE_MARKER
    slides = extract(sys.argv[1], case_marker)
    sys.stdout.reconfigure(encoding="utf-8")
    json.dump(slides, sys.stdout, ensure_ascii=False)


if __name__ == "__main__":
    main()

File: src/Presentation/SlideTextExtractor.php
<?php

declare(strict_types=1);

namespace CasesBot\Presentation;

/**
 * Тонкая обёртка над Python-скриптом python/slide_text_extractor.py (python-pptx):
 * читает текст слайда (заголовок, буллеты) для подсказок при тегировании (§6, §6.2 ТЗ)
 * и определяет, помечен ли слайд как кейс (метка в заметках докладчика, см. $caseMarker).
 */
class SlideTextExtractor
{
    public function __construct(
        private string $pythonBin,
        private string $scriptPath,
        private string $caseMarker = '#кейс#',
    ) {
    }

    /**
     * @return array<int, array{slide_number: int, title: string, text: string, is_case: bool}>
     */
    public function extract(string $pptxPath): array
    {
        $descriptors = [1 => ['pipe', 'w'], 2 => ['pipe', 'w']];
        $process = proc_open(
            [$this->pythonBin, $this->scriptPath, $pptxPath, $this->caseMarker],
            $descriptors,
            $pipes
        );

        if (!is_resource($process)) {
            throw new \RuntimeException('Не удалось запустить slide_text_extractor.py');
        }

        $stdout = stream_get_contents($pipes[1]);
        $stderr = stream_get_contents($pipes[2]);
        fclose($pipes[1]);
        fclose($pipes[2]);
        $exitCode = proc_close($process);

        if ($exitCode !== 0) {
            throw new \RuntimeException("slide_text_extractor.py завершился с ошибкой ({$exitCode}): {$stderr}");
        }

        $decoded = json_decode((string) $stdout, true);
        if (!is_array($decoded)) {
            throw new \RuntimeException("slide_text_extractor.py вернул некорректный JSON: {$stdout}");
        }

        return $decoded;
    }
}
